File: Symmetric Tree.py
Given the root of a binary tree, check whether it is symmetric, i.e., whether the tree is a mirror image of itself.


A binary tree is symmetric if the left subtree is a mirror reflection of the right subtree.
#code here
'''
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
'''
class Solution:
    def isSymmetric(self, root):
        if not root:
            return True

        queue = deque([(root.left, root.right)])

        while queue:
            a, b = queue.popleft()

            if not a and not b:
                continue
            if not a or not b or a.data != b.data:
                return False

            queue.append((a.left, b.right))
            queue.append((a.right, b.left))

        return True

File: Symmetric Tree.java
Given the root of a binary tree, check whether it is symmetric, i.e., whether the tree is a mirror image of itself.
A binary tree is symmetric if the left subtree is a mirror reflection of the right subtree.

#code here

/*
class of the node of the tree is as
class Node{
    int data;
    Node left;
    Node right;
    Node(int data){
        this.data = data;
        left=null;
        right=null;
    }
}

*/
class Solution {
    public static boolean isSymmetric(Node root) {
        // your code here;
        Stack<Node> left = new Stack<>();
        Stack<Node> right = new Stack<>();
        left.push(root.left);
        right.push(root.right);
        while(!left.isEmpty() && !right.isEmpty())
        {
            Node a = left.pop();
            Node b = right.pop();
            
            if(a==null && b==null)
                continue;
                
            if(a==null || b==null || a.data != b.data)
                return false;
                
            left.push(a.left);
            left.push(a.right);
            
            right.push(b.right);
            right.push(b.left);
        }
        return true;
    }
    
}
